<a href="https://colab.research.google.com/github/caplescia/MPM_200_Plescia/blob/main/Session%209/Copy_of_assignment_9_database_management.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Packages and Scientific Computing Libraries

"""
Python Packages Overview:
-------------------------
Packages in Python are collections of modules that provide reusable code and functionality.
They allow developers to organize and share complex code libraries that can be easily imported
and used across different projects.

Key Characteristics of Packages:
- Organized hierarchically
- Contain multiple related modules
- Can be installed using package managers like pip
- Extend Python's core functionality

Three Essential Data Science Libraries:
--------------------------------------
1. Pandas (pd):
   - Primary library for data manipulation and analysis
   - Provides DataFrame: a 2D labeled data structure with columns of potentially different types
   - Key Features:
     * Reading/writing data from various sources (CSV, Excel, SQL databases)
     * Data cleaning and preprocessing
     * Complex data transformations
     * Powerful data analysis and aggregation tools

2. NumPy (np):
   - Fundamental package for scientific computing in Python
   - Introduces the ndarray (n-dimensional array) for efficient numerical operations
   - Key Features:
     * High-performance multidimensional array object
     * Mathematical and statistical functions
     * Linear algebra operations
     * Foundation for scientific and numerical computing
     * Enables fast computations on large datasets

3. Matplotlib (plt):
   - Comprehensive library for creating static, animated, and interactive visualizations
   - Provides MATLAB-like plotting interface
   - Key Features:
     * Line plots, scatter plots, bar charts, histograms
     * Customizable plot styles and layouts
     * Support for multiple plot types and graph formats
     * Integration with NumPy for data visualization

Practical Significance:
----------------------
These libraries form the backbone of data science and scientific computing in Python,
enabling complex data analysis, numerical computations, and insightful visualizations
with minimal code.
"""

In [1]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Read the Excel file
flavivirus_df = pd.read_excel("/content/Flavivirus host associations.xlsx")

# Display basic information about the dataset
print(flavivirus_df.info())

# Show the first few rows of the data
flavivirus_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  570 non-null    int64  
 1   Virus                   570 non-null    object 
 2   Scientific Name         570 non-null    object 
 3   Authors                 570 non-null    object 
 4   Year                    430 non-null    object 
 5   Title/accession number  570 non-null    object 
 6   Journal                 425 non-null    object 
 7   Issue                   415 non-null    float64
 8   Volume                  404 non-null    object 
 9   Pages                   404 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 44.7+ KB
None


sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-time fluorogenic reverse transcription po...   

                     Journal  Issue Volume   Pages  
0  Southwestern Entomologist    NaN    NaN     NaN  
1                        NaN    NaN    NaN     NaN  
2     Zoonoses Public Health   56.0    6-7  338-56  
3     Zoonoses Public Health   56.0    6-7  338-56  
4         J Vet Diagn Invest   24.0      5  959-63

## Using pandas functions that you have seen in the class address following questions
1. How many entries are there in the dataset?
2. How many viruses are present in the dataset?
3. List those viruses.
4. How many rows of data are present for each virus?
5. How many host species are present in the data?
6. Make a list of host species present.
7. Count the number of unique host species for each virus.
8. Which virus has the most number of hosts?
9. Which host has most number of viruses?

In [2]:
# Exploring Virus and Species Diversity in the Dataset
flavivirus_df.columns.tolist()
num_viruses = flavivirus_df['Virus'].nunique()
num_hosts = flavivirus_df['Scientific Name'].nunique()

print(f"Number of unique viruses: {num_viruses}")
print(f"Number of unique host species: {num_hosts}")

# Count total number of entries in the dataset
total_entries = flavivirus_df.shape[0]
print(f"Total number of entries: {total_entries}")


# Count of unique viruses
unique_viruses = flavivirus_df['Virus'].nunique()
print(f"Number of unique viruses: {unique_viruses}")

# List of unique viruses
unique_viruses = flavivirus_df['Virus'].unique()
print (unique_viruses)

# Count of each unique virus
virus_counts = flavivirus_df['Virus'].value_counts()
print(virus_counts)

# Count unique host species
unique_hosts = flavivirus_df['Scientific Name'].nunique()
print(f"Number of unique hosts: {unique_hosts}")

# List of unique host species
unique_hosts = flavivirus_df['Scientific Name'].unique()
print (unique_hosts)

# Count of host species for each virus
virus_host_counts = flavivirus_df.groupby('Virus')['Scientific Name'].nunique()
print(virus_host_counts)

# Most diverse viruses (viruses with most unique host species)
virus_host_counts = flavivirus_df.groupby('Virus')['Scientific Name'].nunique()

virus_host_counts_sorted = virus_host_counts.sort_values(ascending=False)

print(virus_host_counts_sorted.head(5))  # Top 5 viruses with most unique hosts



Number of unique viruses: 30
Number of unique host species: 419
Total number of entries: 570
Number of unique viruses: 30
['Apoi virus' 'Bagaza virus' 'Banzi virus' 'Bouboui virus'
 'Bussuquara virus' 'Cacipacore virus' 'Dakar bat virus' 'Dengue virus'
 'Entebbe bat virus' 'Iguape virus' 'Ilheus virus'
 'Japanese encephalitis virus' 'Koutango virus' 'Kunjin virus'
 'Kyasanur forest virus' 'Louping-ill virus' 'Modoc virus'
 'Murray Valley encephalitis virus' 'Omsk hemorrhagic fever virus'
 'Rio Bravo virus' 'Rocio virus' 'St. Louis Encephalitis virus'
 'Tembusu virus' 'Tickborne encephalitis virus' 'Uganda S virus'
 'Usutu virus' 'Wesselsbron virus' 'West Nile virus' 'Yellow fever virus'
 'Zika virus']
Virus
West Nile virus                     194
Tickborne encephalitis virus         84
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yel

# Metadata

## Given that we are interested in the metadata associated with these host species, we will import datasets associated with the metadata for both bird and mammalian species.

1. Birds

In [3]:
birds_meta = pd.read_excel("/content/Handbook of the Birds of the World and BirdLife International Digital Checklist of the Birds of the World_Version_9.xlsx")
birds_meta.head()


Seq.  Subsp. Seq.             Order    Family name     Family Subfamily  \
0   1.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
1   1.0          1.0               NaN            NaN        NaN       NaN   
2   1.0          2.0               NaN            NaN        NaN       NaN   
3   1.0          3.0               NaN            NaN        NaN       NaN   
4   1.0          4.0               NaN            NaN        NaN       NaN   

  Tribe     Common name             Scientific name       Authority  \
0   NaN  Common Ostrich            Struthio camelus  Linnaeus, 1758   
1   NaN             NaN   Struthio camelus syriacus             NaN   
2   NaN             NaN    Struthio camelus camelus             NaN   
3   NaN             NaN  Struthio camelus massaicus             NaN   
4   NaN             NaN  Struthio camelus australis             NaN   

  2024 IUCN Red List category Synonyms Alternative common names  \
0                          LC      NaN                      NaN   
1                         NaN      NaN                      NaN   
2                         NaN      NaN                      NaN   
3                         NaN      NaN                      NaN   
4                         NaN      NaN                      NaN   

                                 Taxonomic source(s)    SISRecID  \
0  del Hoyo, J., Collar, N.J., Christie, D.A., El...  45020636.0   
1                                                NaN         NaN   
2                                                NaN         NaN   
3                                                NaN         NaN   
4                                                NaN         NaN   

   SpcRecID\n(used before SISRecID)    SubsppID  
0                         1016860.0         NaN  
1                               NaN  45020636_1  
2                               NaN  45020636_2  
3                               NaN  45020636_3  
4                               NaN  45020636_4

In [4]:
birds_meta.shape


(33265, 17)

In [5]:
#species level rows only
species_only = birds_meta[birds_meta["Subsp. Seq."] == 0]

In [6]:
species_count = (birds_meta["Subsp. Seq."] == 0).sum()
print(species_count)



11204


In [7]:
birds_meta[["Seq.", "Subsp. Seq.", "Scientific name"]]




Seq.  Subsp. Seq.             Scientific name
0           1.0          0.0            Struthio camelus
1           1.0          1.0   Struthio camelus syriacus
2           1.0          2.0    Struthio camelus camelus
3           1.0          3.0  Struthio camelus massaicus
4           1.0          4.0  Struthio camelus australis
...         ...          ...                         ...
33260  102032.0          NaN          Tangara cyanoptera
33261  102033.0          NaN            Tangara vassorii
33262  102034.0          NaN               Tangara velia
33263  102035.0          NaN        Tangara brasiliensis
33264  102036.0          NaN           Tangara cucullata

[33265 rows x 3 columns]

In [8]:
#Group subspecies by species
species_groups = birds_meta.groupby("Seq.")


In [9]:
#How many subspecies does each species have
subspecies_counts = birds_meta.groupby("Seq.").size()
print(subspecies_counts)


Seq.
1.0         5
2.0         1
3.0         6
4.0         4
5.0         1
           ..
102032.0    1
102033.0    1
102034.0    1
102035.0    1
102036.0    1
Length: 13232, dtype: int64


The data for birds is at subspecies level. It means that there is species name represented by column `Seq.`. All rows in same number represent one species.

2. Mammals

## Similarly, read file for mammals metadata.

In [10]:
import pandas as pd

mammals_meta = pd.read_csv(
    "/content/IUCN Mammals, Birds, Reptiles, and Amphibians.csv",
    encoding="latin1"
)

mammals_meta.head()


Species ID   Kingdom    Phylum     Class             Order       Family  \
0     42641.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
1  22687170.0  ANIMALIA  CHORDATA      AVES  CAPRIMULGIFORMES  TROCHILIDAE   
2     17879.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
3  47760825.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA   CRICETIDAE   
4     42656.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA  ABROCOMIDAE   

          Genus    Species                  Authority  Infraspecific rank  \
0     Abditomys   latidens            (Sanborn, 1952)                 NaN   
1      Abeillia   abeillei  (Lesson & DeLattre, 1839)                 NaN   
2   Abeomelomys      sevia    (Tate & Archbold, 1935)                 NaN   
3  Abrawayaomys    ruschii         Cunha & Cruz, 1979                 NaN   
4      Abrocoma  bennettii           Waterhouse, 1837                 NaN   

   ...                             Synonyms  \
0  ...                                  NaN   
1  ...                                  NaN   
2  ...  Melomys sevia|Pogonomelomys sevia     
3  ...                                  NaN   
4  ...                                  NaN   

                                  Common names (Eng)  Common names (Fre)  \
0                            Luzon Broad-toothed Rat                 NaN   
1                        Emerald-chinned Hummingbird                 NaN   
2  Menzies' Mouse, Menzies's Mouse, Highland Brus...                 NaN   
3                                       Ruschi's Rat                 NaN   
4                           Bennett's Chinchilla Rat                 NaN   

  Common names (Spa) Red List status Red List criteria  \
0                NaN              DD               NaN   
1                NaN              LC               NaN   
2                NaN              LC               NaN   
3                NaN              LC               NaN   
4                NaN              LC               NaN   

  Red List criteria version Year assessed Population trend  Petitioned  
0                       3.1        2008.0          unknown           N  
1                       3.1        2012.0       decreasing           N  
2                       3.1        2016.0          unknown           N  
3                       3.1        2016.0          unknown           N  
4                       3.1        2016.0          unknown           N  

[5 rows x 23 columns]

In [11]:
mammals_meta.shape

(16081, 23)

In [12]:
filtered_mammals = mammals_meta[mammals_meta['Class'] == 'MAMMALIA']


In [13]:
filtered_mammals

Species ID   Kingdom    Phylum     Class            Order  \
0         42641.0  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   
2         17879.0  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   
3      47760825.0  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   
4         42656.0  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   
5            18.0  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   
...           ...       ...       ...       ...              ...   
16075         NaN  ANIMALIA  CHORDATA  MAMMALIA       CHIROPTERA   
16076         NaN  ANIMALIA  CHORDATA  MAMMALIA  CETARTIODACTYLA   
16078         NaN  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   
16079         NaN  ANIMALIA  CHORDATA  MAMMALIA  CETARTIODACTYLA   
16080    972559.0  ANIMALIA  CHORDATA  MAMMALIA         RODENTIA   

               Family         Genus            Species  \
0             MURIDAE     Abditomys           latidens   
2             MURIDAE   Abeomelomys              sevia   
3          CRICETIDAE  Abrawayaomys            ruschii   
4         ABROCOMIDAE      Abrocoma          bennettii   
5         ABROCOMIDAE      Abrocoma        boliviensis   
...               ...           ...                ...   
16075  PHYLLOSTOMIDAE      Uroderma               spp.   
16076       CAMELIDAE       Vicugna              pacos   
16078         MURIDAE      Apodemus                NaN   
16079          SUIDAE           Sus  scrofa domesticus   
16080         MURIDAE    Hylomyscus              pamfi   

                     Authority  Infraspecific rank  ...  \
0              (Sanborn, 1952)                 NaN  ...   
2      (Tate & Archbold, 1935)                 NaN  ...   
3           Cunha & Cruz, 1979                 NaN  ...   
4             Waterhouse, 1837                 NaN  ...   
5       Glanz & Anderson, 1990                 NaN  ...   
...                        ...                 ...  ...   
16075                      NaN                 NaN  ...   
16076                      NaN                 NaN  ...   
16078                      NaN                 NaN  ...   
16079                      NaN                 NaN  ...   
16080                      NaN                 NaN  ...   

                                  Synonyms  \
0                                      NaN   
2      Melomys sevia|Pogonomelomys sevia     
3                                      NaN   
4                                      NaN   
5                                      NaN   
...                                    ...   
16075                                  NaN   
16076                                  NaN   
16078                                  NaN   
16079                                  NaN   
16080                                  NaN   

                                      Common names (Eng)  Common names (Fre)  \
0                                Luzon Broad-toothed Rat                 NaN   
2      Menzies' Mouse, Menzies's Mouse, Highland Brus...                 NaN   
3                                           Ruschi's Rat                 NaN   
4                               Bennett's Chinchilla Rat                 NaN   
5                                Bolivian Chinchilla Rat                 NaN   
...                                                  ...                 ...   
16075                                                NaN                 NaN   
16076                                                NaN                 NaN   
16078                                                NaN                 NaN   
16079                                                NaN                 NaN   
16080                                                NaN                 NaN   

      Common names (Spa) Red List status Red List criteria  \
0                    NaN              DD               NaN   
2                    NaN              LC               NaN   
3                    NaN              LC               NaN   
4                    N

In [14]:
total_mammal_species = filtered_mammals['Species'].nunique()
print(total_mammal_species)


3929


## For both birds and mammals answer the following questions.
1. Size of the datasets
2. For birds, filter the dataset using species-level rows only. Hint. Column `Subsp. Seq.` value of 0 indicates the main species name
3. How many bird species and mammal species are present in data frames?

1.  For the birds:  the dataset size is (33265, 17).  For the mammals:  the dataset size is (16081, 23).
2.  see code
3.  There are 11,204 bird species.  There are 3929 mammal species.  

# Database Integration Exercise
The main objective of this exercise is to create a comprehensive database by combining three data frames. Follow these steps  n your analysis.

## Create a concept map:  
* Illustrate the relationships between the three data frames.  
* Identify and highlight the Primary Keys in each data frame.  
* Use a diagramming tool (e.g., PowerPoint, LucidChart, or Draw.io) to create a visual representation of these connections.  
* Pay special attention to how you plan to merge the Mammals and Birds data frames.  

## Perform a vertical join of Mammals and Birds data frames:  
* Identify and select only the common columns present in both data frames.  
* Use `pd.concat()` or `df.append()` to create a larger dataframe called **metadata**.  
* Ensure the resulting dataframe has the correct number of rows and columns.  

In [15]:
# Find common columns
common_cols = mammals_meta.columns.intersection(birds_meta.columns)

# Subset both dataframes to only common columns
mammals_common = mammals_meta[common_cols]
birds_common = birds_meta[common_cols]

# Vertically concatenate
metadata = pd.concat([mammals_common, birds_common], ignore_index=True)

# Check shape
print("Rows, Columns:", metadata.shape)
metadata.head()


Rows, Columns: (49346, 4)


Order       Family                  Authority  \
0          RODENTIA      MURIDAE            (Sanborn, 1952)   
1  CAPRIMULGIFORMES  TROCHILIDAE  (Lesson & DeLattre, 1839)   
2          RODENTIA      MURIDAE    (Tate & Archbold, 1935)   
3          RODENTIA   CRICETIDAE         Cunha & Cruz, 1979   
4          RODENTIA  ABROCOMIDAE           Waterhouse, 1837   

                              Synonyms  
0                                  NaN  
1                                  NaN  
2  Melomys sevia|Pogonomelomys sevia    
3                                  NaN  
4                                  NaN

## Merge the metadata dataframe with the Flavivirus dataframe:  
* Experiment with different join types: `inner`, `outer`, `left`, and `right`.  
* For each join type, record the resulting dataframe's shape (rows and columns).  
* Determine the **theoretically correct size of the merged dataframe if done manually**.  
* Compare the results of each join type with the expected outcome.  
* Choose the most appropriate join type based on your analysis.  

In [16]:
# Find common columns between the two dataframes
metadata.columns.intersection(flavivirus_df.columns)


Index([], dtype='object')

In [17]:
metadata.columns


Index(['Order', 'Family', 'Authority', 'Synonyms'], dtype='object')

In [18]:
common_columns_set = set(flavivirus_df.columns).intersection(set(metadata.columns))


In [19]:
print(len(common_columns_set))

0


In [20]:
species_only = birds_meta[birds_meta["Subsp. Seq."] == 0].copy()
birds_species_df = species_only[['Scientific name', 'Order', 'Family']].copy()
birds_species_df.rename(columns={'Scientific name': 'Scientific_Name_Host'}, inplace=True)

filtered_mammals_df = mammals_meta[mammals_meta['Class'] == 'MAMMALIA'].copy()
mammals_species_df = filtered_mammals_df[['Species', 'Order', 'Family']].copy()
mammals_species_df.rename(columns={'Species': 'Scientific_Name_Host'}, inplace=True)

metadata_for_merge = pd.concat([birds_species_df, mammals_species_df], ignore_index=True)

first_merge = pd.merge(
    flavivirus_df,
    metadata_for_merge,
    left_on='Scientific Name',
    right_on='Scientific_Name_Host',
    how='left'
)
print("Shape of the merged dataframe:", first_merge.shape)
first_merge.head()

Shape of the merged dataframe: (570, 13)


sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-time fluorogenic reverse transcription po...   

                     Journal  Issue Volume   Pages Scientific_Name_Host  \
0  Southwestern Entomologist    NaN    NaN     NaN                  NaN   
1                        NaN    NaN    NaN     NaN                  NaN   
2     Zoonoses Public Health   56.0    6-7  338-56                  NaN   
3     Zoonoses Public Health   56.0    6-7  338-56                  NaN   
4         J Vet Diagn Invest   24.0      5  959-63       Alectoris rufa   

         Order                                  Family  
0          NaN                                     NaN  
1          NaN                                     NaN  
2          NaN                                     NaN  
3          NaN                                     NaN  
4  GALLIFORMES  Pheasants, Partridges, Turkeys, Grouse

In [21]:
first_merge.shape

(570, 13)

In [22]:
outer_merge = pd.merge (
    flavivirus_df,
    metadata_for_merge,
    left_on='Scientific Name',
    right_on='Scientific_Name_Host',
    how='outer'
)

In [23]:
outer_merge.shape

(17161, 13)

In [24]:
outer_merge.head()

sr no. Virus Scientific Name Authors Year Title/accession number Journal  \
0     NaN   NaN             NaN     NaN  NaN                    NaN     NaN   
1     NaN   NaN             NaN     NaN  NaN                    NaN     NaN   
2     NaN   NaN             NaN     NaN  NaN                    NaN     NaN   
3     NaN   NaN             NaN     NaN  NaN                    NaN     NaN   
4     NaN   NaN             NaN     NaN  NaN                    NaN     NaN   

   Issue Volume Pages      Scientific_Name_Host             Order  \
0    NaN    NaN   NaN         Abeillia abeillei  CAPRIMULGIFORMES   
1    NaN    NaN   NaN    Abroscopus albogularis     PASSERIFORMES   
2    NaN    NaN   NaN    Abroscopus schisticeps     PASSERIFORMES   
3    NaN    NaN   NaN  Abroscopus superciliaris     PASSERIFORMES   
4    NaN    NaN   NaN            Aburria aburri       GALLIFORMES   

                          Family  
0                   Hummingbirds  
1                  Bush-warblers  
2                  Bush-warblers  
3                  Bush-warblers  
4  Guans, Chachalacas, Curassows

### Create a pivote table,
1. Rows are virueses  
2. Columns are Taxonomic orders  
3. Values are number of unique host species  

In [25]:
pivot = pd.pivot_table(
    first_merge,
    index="Virus",          # rows
    columns="Order",        # columns
    values="Scientific_Name_Host",  # column with host names (corrected from 'Host_Species')
    aggfunc=pd.Series.nunique,
    fill_value=0
)

pivot

Order                             ACCIPITRIFORMES  ANSERIFORMES  \
Virus                                                             
Bagaza virus                                    0             0   
Ilheus virus                                    0             0   
Japanese encephalitis virus                     0             0   
Koutango virus                                  0             0   
Kunjin virus                                    0             0   
Louping-ill virus                               0             0   
Murray Valley encephalitis virus                0             0   
Rocio virus                                     0             0   
St. Louis Encephalitis virus                    1             0   
Tembusu virus                                   0             1   
Tickborne encephalitis virus                    1             7   
Uganda S virus                                  0             0   
Usutu virus                                     4             0   
West Nile virus                                16             4   

Order                             BUCEROTIFORMES  CAPRIMULGIFORMES  \
Virus                                                                
Bagaza virus                                   0                 0   
Ilheus virus                                   0                 0   
Japanese encephalitis virus                    0                 0   
Koutango virus                                 0                 0   
Kunjin virus                                   0                 0   
Louping-ill virus                              0                 0   
Murray Valley encephalitis virus               0                 0   
Rocio virus                                    0                 0   
St. Louis Encephalitis virus                   0                 0   
Tembusu virus                                  0                 0   
Tickborne encephalitis virus                   0                 0   
Uganda S virus                                 0                 0   
Usutu virus                                    1                 1   
West Nile virus                                1                 0   

Order                             CATHARTIFORMES  CHARADRIIFORMES  \
Virus                                                               
Bagaza virus                                   0                0   
Ilheus virus                                   0                0   
Japanese encephalitis virus                    0                0   
Koutango virus                                 0                0   
Kunjin virus                                   0                0   
Louping-ill virus                              0                0   
Murray Valley encephalitis virus               0                0   
Rocio virus                                    0                0   
St. Louis Encephalitis virus                   0                1   
Tembusu virus                                  0                0   
Tickborne encephalitis virus                   0                2   
Uganda S virus                                 0                0   
Usutu virus                                    0                3   
West Nile virus                                2                9   

Order                             CICONIIFORMES  COLUMBIFORMES  CORACIIFORMES  \
Virus                                                                           
Bagaza virus                                  0              0              0   
Ilheus virus                                  0              0              0   
Japanese encephalitis virus                   0              2              0   
Koutango virus                                0              0              0   
Kunjin virus                                  0              0              0   
Louping-ill virus                             0              0              0   
Murray Valley encephalitis virus              0      

In [26]:
pivot = pd.pivot_table(
    outer_merge,
    index="Virus",          # rows
    columns="Order",        # columns
    values="Scientific_Name_Host",  # column with host names (corrected from 'Host_Species')
    aggfunc=pd.Series.nunique,
    fill_value=0
)

pivot

Order                             ACCIPITRIFORMES  ANSERIFORMES  \
Virus                                                             
Bagaza virus                                    0             0   
Ilheus virus                                    0             0   
Japanese encephalitis virus                     0             0   
Koutango virus                                  0             0   
Kunjin virus                                    0             0   
Louping-ill virus                               0             0   
Murray Valley encephalitis virus                0             0   
Rocio virus                                     0             0   
St. Louis Encephalitis virus                    1             0   
Tembusu virus                                   0             1   
Tickborne encephalitis virus                    1             7   
Uganda S virus                                  0             0   
Usutu virus                                     4             0   
West Nile virus                                16             4   

Order                             BUCEROTIFORMES  CAPRIMULGIFORMES  \
Virus                                                                
Bagaza virus                                   0                 0   
Ilheus virus                                   0                 0   
Japanese encephalitis virus                    0                 0   
Koutango virus                                 0                 0   
Kunjin virus                                   0                 0   
Louping-ill virus                              0                 0   
Murray Valley encephalitis virus               0                 0   
Rocio virus                                    0                 0   
St. Louis Encephalitis virus                   0                 0   
Tembusu virus                                  0                 0   
Tickborne encephalitis virus                   0                 0   
Uganda S virus                                 0                 0   
Usutu virus                                    1                 1   
West Nile virus                                1                 0   

Order                             CATHARTIFORMES  CHARADRIIFORMES  \
Virus                                                               
Bagaza virus                                   0                0   
Ilheus virus                                   0                0   
Japanese encephalitis virus                    0                0   
Koutango virus                                 0                0   
Kunjin virus                                   0                0   
Louping-ill virus                              0                0   
Murray Valley encephalitis virus               0                0   
Rocio virus                                    0                0   
St. Louis Encephalitis virus                   0                1   
Tembusu virus                                  0                0   
Tickborne encephalitis virus                   0                2   
Uganda S virus                                 0                0   
Usutu virus                                    0                3   
West Nile virus                                2                9   

Order                             CICONIIFORMES  COLUMBIFORMES  CORACIIFORMES  \
Virus                                                                           
Bagaza virus                                  0              0              0   
Ilheus virus                                  0              0              0   
Japanese encephalitis virus                   0              2              0   
Koutango virus                                0              0              0   
Kunjin virus                                  0              0              0   
Louping-ill virus                             0              0              0   
Murray Valley encephalitis virus              0      